pkgs/development/compilers/graalvm/community-edition/default.nix

```diff
@@ -1,17 +1,62 @@
-{ lib, callPackage, Foundation }:
+{ callPackage, Foundation }:
+/*
+  Add new graal versions and products here and then see update.nix on how to
+  generate the sources.
+*/
 
 let
   mkGraal = opts: callPackage (import ./mkGraal.nix opts) {
     inherit Foundation;
   };
+
+  commonProducts = [
+    "graalvm-ce"
+    "native-image-installable-svm"
+    "ruby-installable-svm"
+    "wasm-installable-svm"
+  ];
+
+  /*
+    Looks a bit ugly but makes version update in the update script using sed
+    much easier
+  */
+  graalvm11-ce-release-version = "22.0.0.2";
+  graalvm17-ce-release-version = "22.0.0.2";
+  graalvm11-ce-dev-version = "22.2.0-dev-20220331_1955";
+  graalvm17-ce-dev-version = "22.2.0-dev-20220331_1955";
+
 in
 {
   inherit mkGraal;
 
   graalvm11-ce = mkGraal rec {
-    version = "22.0.0.2";
+    config = {
+      x86_64-darwin = {
+        arch = "darwin-amd64";
+        products = commonProducts ++ [ "python-installable-svm" ];
+        version = graalvm11-ce-release-version;
```

Comment:
Maybe we can pass this as a parameter on `mkGraal`, and if someone overrides it in the config we them use this version.

This way, we don't need to explicitly pass the version for every other platform (just `aarch64-linux`).

Reply:
Doing it this way we wouldn't even need those constants. We can just use the version directly because there wouldn't be repetitions.

Reply:
I am not sure if I understand you completely. I have extracted the commonConfig to a `baseConfig` to reduce the repetition in the latest commit. If you meant something else, could you please elaborate?

Reply:
I was thinking more of something like this. On `mkGraal.nix`, you can add one more parameter:

```nix
{ javaVersion
, platforms
, config
, defaultVersion
, useMusl ? false
}:
```

Then you could do something like this:

```nix
graalvm17-ce = mkGraal rec {
  defaultVersion = "22.0.0.2";
  config = {
    aarch64-darwin = {
      arch = "darwin-aarch64";
      products = [
        "graalvm-ce"
        "native-image-installable-svm"
      ];
      version = "22.2.0-dev-20220331_1955";
    };
  };
};
```

This way there is no repetition, and it is very clear what is happening IMO.

Reply:
That is what we had before and I tried that but this leads to a problem in the update script when creating the
`product|javaVersion|graalVersion` for each platform.

Since we have now two possible versions, dev and release depending on the platform, I need to specify them per platform. Otherwise the defaultVersion overwrites the dev version in the `graalvmXX-ce-sources.json` file or vice versa.

That was one of the reasons why I first added a new `graalvmXX-ce-dev` name because then I had only one version per derivation, eg. dev for `graalvmXX-ce-dev` and release for `graalvmXX-ce` so I didn't have to distinguish.

I don't like the current repetition either but we need it temporarily. Since we are planning to remove the dev versions when `aarch64-darwin` makes it into the release, we can then remove this repetition too.

Reply:
Maybe I am missing something, but not sure why the proposed solution wouldn't work. You can use `config.version` if it exists, or use `defaultVersion` if it doesn't (of course, you need pass `defaultVersion` to the update script as an argument too). Something like this:

```
nix-repl> {}.version or "22.0.0.5"
"22.0.0.5"
```

Where `{}` would be the config attrset.

Reply:
Ah now I get it:smiley:. Sure that works. Let me quickly implement that.

Reply:
Ok, I have added the `defaultVersion` but went back to a config block for `graalvm11-ce` and `graalvm17-ce` because I like to explicitly configure it there instead of creating a baseConfig set.